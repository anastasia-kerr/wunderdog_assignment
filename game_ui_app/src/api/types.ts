import { Gestures } from '@/types'

type BaseApiResponse<T> = {
    succeeded: boolean;
    errors: any[];
    result: T;
}
export type CreateGameResponse = BaseApiResponse<{
    id: number
}>
export type JoinCurrentRoundResponse = BaseApiResponse<{
    id: number;
    roundNumber: number
}>
export type GetCurrentRoundPlayersResponse = BaseApiResponse<{
    players: string[]
}>
export type RoundResult = {
    isDraw?: boolean;
    isLastRound?: boolean;
    loserGesture?: Gestures;
    loserNickname?: string;
    winnerGesture?: Gestures;
    winnerNickname?: string;
}
export type GetRoundResultResponse = BaseApiResponse<RoundResult>
export type GetGameResultResponse = BaseApiResponse<
{winnerNickname?: string}
>
export type LeaveCurrentRoundResponse = JoinCurrentRoundResponse
export type PlaceGestureResponse = JoinCurrentRoundResponse
