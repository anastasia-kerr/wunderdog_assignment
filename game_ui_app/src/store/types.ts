import { RoundResult } from '@/api/types'

export interface RootState {
    error: string,
    gameId: string,
    roundNumber: number,
    currentPlayers: string[],
    gameWinner: string,
    nickname?: string|null,
    playerIdentifier?: string|null,
    roundResults: RoundResult
  }
