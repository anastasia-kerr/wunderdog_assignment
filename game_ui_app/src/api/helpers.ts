import config from '../configs'

export const poll = async (callback:() => Promise<any>) => {
  let pollCount = 0
  const handler = async () => {
    pollCount++
    const result = await callback()
    if (result || pollCount === config.maxPollingTimeInSeconds) {
      clearInterval(intervalId)
    }
  }
  const intervalId = setInterval(handler, config.pollingIntervalInMs)
}
