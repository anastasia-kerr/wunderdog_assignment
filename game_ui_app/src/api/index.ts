import axios from 'axios'
import { CreateGameResponse, GetCurrentRoundPlayersResponse, GetGameResultResponse, GetRoundResultResponse, JoinCurrentRoundResponse, LeaveCurrentRoundResponse, PlaceGestureResponse } from './types'

function setUpInterceptors () {
  axios.interceptors.response.use(null, (error) => {
    if (error?.response?.data?.Errors) {
      console.log('Error while calling the api: ', JSON.stringify(error?.response?.data?.Errors))
    } else {
      console.log('unexpected error while calling the api: ', error)
    }
    return Promise.reject(error)
  })
}
export function setApiDefaults (url:string) {
  axios.defaults.baseURL = url
  setUpInterceptors()
}
export async function createGame (payload : {nickname:string, numberOfRounds:number}):Promise<number> {
  const { data } = await axios.post<CreateGameResponse>(
    'api/Games',
    payload
  )
  return data.result.id
}
export async function joinCurrentRound (payload : { gameId:string, playerNickname:string, playerIdentifier:string}):Promise<number> {
  const { data } = await axios.put<JoinCurrentRoundResponse>(
    `api/Games/${payload.gameId}/join`,
    payload
  )
  return data.result.roundNumber
}
export async function getCurrentRoundPlayers (gameId:string):Promise<string[]> {
  const { data } = await axios.get<GetCurrentRoundPlayersResponse>(
    `api/Games/${gameId}/players`
  )
  return data.result.players
}
export async function getRoundResults (gameId:string, round:number):Promise<GetRoundResultResponse> {
  const { data } = await axios.get<GetRoundResultResponse>(
    `api/Games/${gameId}/result/${round}`
  )
  return data
}
export async function getGameResult (gameId:string):Promise<GetGameResultResponse> {
  const { data } = await axios.get<GetGameResultResponse>(
    `api/Games/${gameId}/result/`
  )
  return data
}
export async function leaveCurrentRound (payload : { gameId:string, playerIdentifier:string}):Promise<number> {
  const { data } = await axios.put<LeaveCurrentRoundResponse>(
    `api/Games/${payload.gameId}/leave`,
    payload
  )
  return data.result.roundNumber
}
export async function placeGesture (payload : { gameId:string, playerNickname:string, playerIdentifier:string, roundGesture:number}):Promise<number> {
  const { data } = await axios.post<PlaceGestureResponse>(
    `api/Games/${payload.gameId}/gesture`,
    payload
  )
  return data.result.id
}
