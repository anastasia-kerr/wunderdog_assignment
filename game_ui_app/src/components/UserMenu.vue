<template>
  <div class="text-center">
    <v-menu offset-y>
      <template v-slot:activator="{ on, attrs }">
        <v-btn color="secondary" text v-bind="attrs" v-on="on" v-text="nickname" />
      </template>
      <v-list>
        <v-list-item>
          <v-btn id="leave_button" text color="primary" @click="signOut">{{ signOutLabel }}</v-btn>
        </v-list-item>
      </v-list>
    </v-menu>
    <nickname-dialog id="nickname_dialog" v-if="showNicknameDialog" />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import NicknameDialog from './dialogs/NicknameDialog.vue'
import { mapState } from 'vuex'

export default Vue.extend({
  name: 'UserMenu',
  components: { NicknameDialog },
  data: () => ({
    signOutLabel: 'Leave game'
  }),
  computed: {
    ...mapState([
      'nickname'
    ]),
    showNicknameDialog (): boolean {
      return !this.nickname
    }
  },
  methods: {
    async signOut (): Promise<void> {
      await this.$store.dispatch('leaveGame')
      this.$router.push({ name: 'home' })
    }
  }
})
</script>
