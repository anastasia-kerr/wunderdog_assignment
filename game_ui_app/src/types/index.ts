export enum Gestures {
    Rock = 0,
    Paper = 1,
    Scissors = 2
  }
