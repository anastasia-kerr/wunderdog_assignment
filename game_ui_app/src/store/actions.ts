import { ActionTree } from 'vuex'
import { RootState } from './types'
import { leaveCurrentRound, joinCurrentRound, getCurrentRoundPlayers, getRoundResults, getGameResult, placeGesture } from '@/api'
import { poll } from '@/api/helpers'

const actions: ActionTree<RootState, RootState> = {
  setUser (context, nickname: string) {
    context.commit('setNickname', nickname)
    localStorage.setItem('nickname', nickname ?? '')
    localStorage.setItem('playerIdentifier', nickname ? nickname + Date.now().toString() : '')
  },
  async leaveGame ({ state, dispatch }) {
    if (state.playerIdentifier) {
      await leaveCurrentRound({ gameId: state.gameId, playerIdentifier: state.playerIdentifier })
    }
    dispatch('setUser', '')
  },
  async joinCurrentRound ({ state, commit, dispatch }) {
    if (!state.playerIdentifier || !state.nickname) {
      throw new Error('Player must have some identifier or name')
    }

    try {
      const roundNumber = await joinCurrentRound({
        gameId: state.gameId,
        playerNickname: state.nickname,
        playerIdentifier: state.playerIdentifier
      })

      commit('resetGameState')
      commit('setRoundNumber', roundNumber)

      dispatch('setCurrentPlayers')
    } catch (error : any) {
      commit('setError', error?.response?.data?.Errors)
    }
  },
  async setCurrentPlayers ({ commit, state }) {
    return await poll(async () => {
      const result = await getCurrentRoundPlayers(state.gameId)
      if (result.length === 2) {
        commit('setCurrentPlayers', result)
        return true
      } else if (state.currentPlayers.length === 2) {
        return true
      }
      return false
    })
  },
  async setCurrentRoundResults ({ state, commit, dispatch }) {
    return await poll(async () => {
      const response = await getRoundResults(state.gameId, state.roundNumber)
      if (response.succeeded) {
        commit('setCurrentRoundResults', response.result)
        if (response.result.isLastRound) {
          dispatch('setGameResult')
        }
        return true
      }
      return false
    })
  },
  async setGameResult ({ commit, state }) {
    const response = await getGameResult(state.gameId)
    commit('setGameResult', response.result.winnerNickname)
  },
  async placeGesture (context, payload: { gesture: number }) {
    if (context.state.playerIdentifier && context.state.nickname) {
      await placeGesture({
        gameId: context.state.gameId,
        playerNickname: context.state.nickname,
        playerIdentifier: context.state.playerIdentifier,
        roundGesture: payload.gesture
      })
    }
  }
}
export default actions
