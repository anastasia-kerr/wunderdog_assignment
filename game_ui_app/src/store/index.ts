import { RoundResult } from '@/api/types'
import Vue from 'vue'
import Vuex from 'vuex'
import { RootState } from './types'
import getters from './getters'
import mutations from './mutations'
import actions from './actions'

Vue.use(Vuex)
export const state: RootState = {
  gameId: '',
  error: '',
  roundNumber: 0,
  currentPlayers: [] as string[],
  gameWinner: '',
  nickname: localStorage.getItem('nickname'),
  playerIdentifier: localStorage.getItem('playerIdentifier'),
  roundResults: {} as RoundResult
}

export default new Vuex.Store({
  state,
  getters,
  mutations,
  actions
})
