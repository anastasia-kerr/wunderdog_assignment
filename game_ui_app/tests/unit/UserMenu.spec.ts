import Vue from 'vue'
import UserMenu from '@/components/UserMenu.vue'
import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils'
import Vuetify from 'vuetify'
import { RootState } from '@/store/types'
import { RoundResult } from '@/api/types'
import Vuex, { Store } from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)
Vue.use(Vuetify)

describe('UserMenu', () => {
  let vuetify: Vuetify
  let wrapper:Wrapper<Vue>
  let store: Store<RootState>
  const player1 = 'player1'
  const state = {
    gameId: '1',
    error: '',
    nickname: player1,
    roundNumber: 0,
    currentPlayers: [],
    gameWinner: '',
    roundResults: {} as RoundResult
  }
  beforeEach(() => {
    vuetify = new Vuetify()
    store = new Vuex.Store({
      state
    })
    wrapper = shallowMount(UserMenu, { localVue, vuetify, store })
  })
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('Must display player name', () => {
    expect(wrapper.text()).not.toContain(player1)
    const dialog = wrapper.find('#nickname_dialog')
    expect(dialog.exists()).toBeFalsy()
  })

  it('Must have a way to leave the game', () => {
    const btn = wrapper.find('#leave_button')
    expect(btn.exists()).toBeTruthy()
  })

  it('If nickname is not provided must ask nickname before continuing', () => {
    store = new Vuex.Store({
      state: { ...state, nickname: '' }
    })
    wrapper = mount(UserMenu, { localVue, vuetify, store })
    const dialog = wrapper.find('#nickname_dialog')
    expect(dialog.exists()).toBe(true)
  })
})
