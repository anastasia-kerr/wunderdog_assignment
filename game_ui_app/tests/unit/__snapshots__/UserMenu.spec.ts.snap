// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu renders correctly 1`] = `
<div
  class="text-center"
>
  <v-menu-stub
    closedelay="0"
    closeonclick="true"
    closeoncontentclick="true"
    contentclass=""
    maxheight="auto"
    maxwidth="auto"
    nudgebottom="0"
    nudgeleft="0"
    nudgeright="0"
    nudgetop="0"
    nudgewidth="0"
    offsety="true"
    opendelay="0"
    openonclick="true"
    origin="top left"
    transition="v-menu-transition"
  >
     
    <v-list-stub
      tag="div"
    >
      <v-list-item-stub
        activeclass=""
        tag="div"
      >
        <v-btn-stub
          activeclass=""
          color="primary"
          id="leave_button"
          tag="button"
          text="true"
          type="button"
        >
          Leave game
        </v-btn-stub>
      </v-list-item-stub>
    </v-list-stub>
  </v-menu-stub>
   
  <!---->
</div>
`;
